Add documentation to handle (physical) units of recording channels (#3844)
SpikeInterface/spikeinterface: doc/how_to/physical_units.rst

Working with physical units in SpikeInterface recordings
========================================================

> In neurophysiology recordings, data is often stored in raw ADC (Analog-to-Digital Converter) units but needs to be analyzed in physical units.

Reviewer: ```suggestion
In neurophysiology recordings, data is often stored in raw ADC (Analog-to-Digital Converter) units but at various points in an analysis pipeline the end-user may want to observe their data in physical units.
```

I think this line saying "needs to be analyzed" sets us up for my concern for beginners. Because we are basically saying you have to do this. Then the next lines says it's optional. And that is confusing. So maybe we write saying that the user can switch to physical units at anytime they want with this tooling. Then we have the rest of your notes. 

What do you think?

Author: Your proposal is better than what we have. I would still use analyze instead of observe in your assertion but let's move forward with this as it is.

For extracellular recordings, this is typically microvolts (µV), but some recording devices may use different physical units.
SpikeInterface provides tools to handle both situations.

It's important to note that **most spike sorters work fine on raw digital (ADC) units** and not scaling is needed.
Many preprocessing tools are also linear transformations, and if the ADC is implemented as a linear transformation, the overall effect can be preserved.
That is, **preprocessing steps can often be applied either before or after unit conversion without affecting the outcome.**

Therefore, **it is usually safe to work in raw ADC units unless a specific tool or analysis requires physical units**.
If you are interested in visualizations, comparability across devices, or outputs with interpretable physical scales (e.g., microvolts), converting to physical units is recommended.
Otherwise, remaining in raw units can simplify processing and preserve performance.

> 

Reviewer: We might want Alessio or Sam to comment on some of the internal tooling. I think the scaling is automatic for some of our stuff so we should make it clear if/when we do that. I just don't remember off the top of my head. If Alessio doesn't have the time to look this over I can doublecheck in the code.

Author: I think this is only missing part to move this forward?

Reviewer: Yeah I agree @alejoe91 could you comment here when you have a moment :)

Understanding Physical Units
----------------------------

Most recording devices store data in ADC units (integers) to save space and preserve the raw data.
To convert these values to physical units, two parameters are needed:

* **gain**: A multiplicative factor to scale the raw values
* **offset**: An additive factor to shift the values

The conversion formula is:

.. code-block:: text

    physical_value = raw_value * gain + offset

> 

Author: Can you add a suggestion?

> 

Reviewer: Yep will do. Have some experiments all day today, but my hope is tomorrow should be a little freer.

Converting to Physical Units
----------------------------

SpikeInterface provides two preprocessing classes for converting recordings to physical units. Both wrap the
``RecordingExtractor`` class and ensures that the data is returned in physical units when calling `get_traces <https://spikeinterface.readthedocs.io/en/stable/api.html#spikeinterface.core.BaseRecording.get_traces>`_

1. ``scale_to_uV``: The primary function for extracellular recordings. SpikeInterface is centered around
   extracellular recordings, and this function is designed to convert the data to microvolts (µV). Many plotting

>    and analyzing functions in SpikeInterface expect data in microvolts, so this is the recommended approach for most users.

Reviewer: The issue is most plotting functions don't need this because we offer a `return_scaled` so maybe we change this statement. 

If you want to implement your own plotting functions you'll need to make sure to display units in microvolts so you can use this function. You can also use this for SpikeInterace plotting functions although plotting functions in spikeinterface also offer automatically scaling using the `return_scaled` argument

Author: Yes, this statement is misleading, I agree. Let me take a look.

Author: I just removed it.

   physical unit, not just microvolts. This is useful for other types of recordings, such as force measurements in Newtons but should be
   handled with care.

For most users working with extracellular recordings, ``scale_to_uV`` is the recommended choice:

.. code-block:: python

    from spikeinterface.extractors import read_intan
    from spikeinterface.preprocessing import scale_to_uV

    # Load recording (data is in ADC units)
    recording = read_intan("path/to/file.rhs")

    # Convert to microvolts
    recording_uv = scale_to_uV(recording)

For recordings with non-standard units (e.g., force measurements in Newtons), use ``scale_to_physical_units``:

.. code-block:: python

    from spikeinterface.preprocessing import scale_to_physical_units

    # Convert to physical units (whatever they may be)
    recording_physical = scale_to_physical_units(recording)

Both preprocessors automatically:

1. Detect the appropriate gain and offset from the recording properties
2. Apply the conversion to all channels
3. Update the recording properties to reflect that data is now in physical units

Setting Custom Physical Units

> ---------------------------

Reviewer: ----- need to be added :)

Author: Thanks!

Author: Changed them.

While most extractors automatically set the appropriate ``gain_to_uV`` and ``offset_to_uV`` values,
there might be cases where you want to set custom physical units. In these cases, you can set
the following properties:

* ``physical_unit``: The target physical unit (e.g., 'uV', 'mV', 'N')
* ``gain_to_unit``: The gain to convert from raw values to the target unit
* ``offset_to_unit``: The offset to convert from raw values to the target unit

You need to set these properties for every channel, which allows for the case when there are different gains and offsets on different channels. Here's an example:

.. code-block:: python

    # Set custom physical units
    num_channels = recording.get_num_channels()
    values = ["volts"] * num_channels
    recording.set_property(key='physical_unit', value=values)

>     values = [0.001] * num_channels  # Convert from ADC to volts

Reviewer: Maybe we say
```
gain_values = [0.001] * num_channels
```
to be even more clear and then below we would say
`offset_values`

just so we aren't using the same variable being overwritten for both? This is definitely optional.

Author: Done.

Working with physical units in SpikeInterface recordings
========================================================

In neurophysiology recordings, data is often stored in raw ADC (Analog-to-Digital Converter) integer values but needs to be analyzed in physical units.
For extracellular recordings, this is typically microvolts (µV), but some recording devices may use different physical units.
SpikeInterface provides tools to handle both situations.

It's important to note that **most spike sorters work fine on raw digital (ADC) units** and scaling is not needed. Going a step further, some sorters, such as Kilosort 3, require their input to be in raw ADC units.

> The specific behavior however depends on the spike sorter, so it's important to understand the specific input requirements in a case per case basis.

Reviewer: ```suggestion
The specific behavior however depends on the spike sorter, so it's important to understand the specific input requirements on a case per case basis.
```
:)

Many preprocessing tools are also linear transformations, and if the ADC is implemented as a linear transformation which is fairly common, then the overall effect can be preserved.

> That is, **preprocessing steps can often be applied either before or after unit conversion without affecting the outcome.**. That said, there are rough edges to this approach.

Reviewer: ```suggestion
That is, **preprocessing steps can often be applied either before or after unit conversion without affecting the outcome.**. That being said, there are rough edges to this approach.
```
this is the way to say that

> Working with Units in SpikeInterface Recordings

Reviewer 2: "Units" could refer to sorted units. Maybe
```suggestion
Working with physical units in SpikeInterface recordings
```

> ===============================================

Reviewer: rendering is impossible for the suggestion but this needs to go to the end of the heading.

Author: Generated by copilot before.
Fixed.

In neurophysiology recordings, data is often stored in raw ADC (Analog-to-Digital Converter) units
but needs to be analyzed in physical units. For extracellular recordings, this is typically microvolts (µV),
but some recording devices may use different physical units. SpikeInterface provides tools to handle both
situations.

Understanding Physical Units
----------------------------

Most recording devices store data in ADC units (integers) to save space and preserve the raw data.
To convert these values to physical units, two parameters are needed:

* **gain**: A multiplicative factor to scale the raw values
* **offset**: An additive factor to shift the values

The conversion formula is:

.. code-block:: text

    physical_value = raw_value * gain + offset

Converting to Physical Units

> -------------------------

Author: Thanks, fixed.

SpikeInterface provides two preprocessing classes for converting recordings to physical units. Both wrap the

> ``RecordingExtractor`` class and ensures that the data is returned in physical units when calling get_traces()

Reviewer 2: I was gonna suggest we link to some docs on `get_traces`, but we don't really have any... maybe we could link to the API
```suggestion
``RecordingExtractor`` class and ensures that the data is returned in physical units when calling `get_traces <https://spikeinterface.readthedocs.io/en/stable/api.html#spikeinterface.core.BaseRecording.get_traces>`_
```

Author: Thanks for the suggestions!

1. ``scale_to_uV``: The primary function for extracellular recordings. SpikeInterface is centered around

>    extracellular recordings, and this function is designed to convert the data to microvolts (µV). Many plottinf

Reviewer 2: spelling
```suggestion
   extracellular recordings, and this function is designed to convert the data to microvolts (µV). Many plotting
```

    recording = read_intan("path/to/file.rhs")

    # Convert to microvolts
    recording_uv = scale_to_uV(recording)

For recordings with non-standard units (e.g., force measurements in Newtons), use ``scale_to_physical_units``:

.. code-block:: python

    from spikeinterface.preprocessing import scale_to_physical_units

    # Convert to physical units (whatever they may be)
    recording_physical = scale_to_physical_units(recording)

Both preprocessors automatically:

1. Detect the appropriate gain and offset from the recording properties
2. Apply the conversion to all channels
3. Update the recording properties to reflect that data is now in physical units

Setting Custom Physical Units
---------------------------

While most extractors automatically set the appropriate ``gain_to_uV`` and ``offset_to_uV`` values,
there might be cases where you want to set custom physical units. In these cases, you can set
the following properties:

* ``physical_unit``: The target physical unit (e.g., 'uV', 'mV', 'N')
* ``gain_to_unit``: The gain to convert from raw values to the target unit
* ``offset_to_unit``: The offset to convert from raw values to the target unit

> Here's an example of how to set custom units:

Reviewer 2: I think the code below makes more sense when you know that you have to set the units for every channel. Could we add something like:

```
You need to set these properties for every channel, which allows for the case when there are different gains and offsets on different channels. Here's an example:
```

Author: Yes, this makes sense.

Author: Changed.
